Add hot reload flag

CohenArthur opened the pull request:
Add the `--online-change` option to enable mangling of functions, as well as globals once that bit is merged upstream. I'm guessing we might want to integrate this with the `got-layout-file` option you're adding in #6 but we can think about this later on

CohenArthur committed:
codegen: Add mangling for global variables

CohenArthur committed:
rusty: Update testsuite for new global variable section names

CohenArthur committed:
section_mangler: Encode structs and enums

CohenArthur committed:
codegen: Mangle struct and enum types properly

CohenArthur committed:
section-mangler: Add base for handling arrays

CohenArthur committed:
codegen: Handle more complex types in mangle_type

CohenArthur committed:
codegen: Mangle type aliases properly

CohenArthur committed:
rusty: Accept snapshots with new section names

CohenArthur committed:
section_mangler: Add base for decoding mangled names

CohenArthur committed:
section-mangler: Add prefix constant, derive Debug and PartialEq

CohenArthur committed:
section_mangler: Add struct decoding

CohenArthur committed:
section_mangler: Add decoding of enum types

CohenArthur committed:
section_mangler: Derive Clone on SectionMangler

CohenArthur committed:
section_mangler: Add .name() method

CohenArthur committed:
section_mangler: Add parsing of mangled functions

lewis-revill committed:
feat: Generate a custom GOT array and save/load its layout

This commit contains changes that are required to generate a global
variable which will be used to store an array of addresses of other
global variables in the program. The location of the globals within this
array should be stable after a recompilation to support online change,
so we support loading a pre-existing layout to ensure that for all
globals that we see in both the current program and the pre-existing
layout, their positions remain the same. Otherwise, any new globals will
be placed in any empty spaces left by old globals, or appended on to the
end of the array. The layout will then be saved back the the file used
for saving and loading.

Currently, to use this feature the flag `--got-layout-file=<file>` must
be provided, which should specify the name of either a TOML or JSON file
to use to save, and optionally load if the file already exists, the GOT
layout. In future we will integrate this with a generic online change
flag, whereby it will not be necessary to ask for a GOT layout file when
we already know that we need it for online change.

CohenArthur committed:
section_mangler: Change API for `SectionMangler::with_return_type`

CohenArthur committed:
codegen: Propagate errors better when creating section names

CohenArthur committed:
rusty: Update insta testsuite for section mangling.

CohenArthur committed:
rusty: Add --online-change CLI flag

CohenArthur commented on compiler/plc_driver/src/pipelines.rs:
```diff
@@ -284,8 +284,16 @@ impl<T: SourceContainer + Sync> AnnotatedProject<T> {
             dependencies,
             &self.index,
+            compile_options.online_change,
         )?;
-        code_generator.generate(context, unit, &self.annotations, &self.index, &llvm_index)
+        code_generator.generate(
+            context,
+            unit,
+            &self.annotations,
+            &self.index,
+            &llvm_index,
+            compile_options.online_change,
+        )
```
I'm not a fan of duplicating the parameter here, but they do the same for `self.index` and it would require the introduction of one more "context-like" type to `code_generator`. Do you think it's worth it?

lewis-revill replied:
Hmm, I do feel like it might fit better as a parameter in the constructor, but I can't articulate precisely why! Online change does feel like a piece of information relevant to codegen at the same granularity as the optimization level or the debug level.

lewis-revill commented:
This looks good to me, my suggestion for how we deal with the `--got-layout-file` argument in conjunction with this is to have a default generated filename based on the output executable name when `--online-change` is provided, and just use `--got-layout-file` to be able to override that if needed. And if it's provided when `--online-change` is not provided, just error.

ghaith committed:
ci: rename workflow, update readme

lewis-revill commented:
Current state looks good to me!

riederm committed:
refactor(AST): introduce AstVisitor trait (#1231)

* refactor(AST): introduce AstVisitor trait

The AST-Visitor trait allows generic visiting of AST-nodes. A default
Walking behavior is implemented for each AstStatement but it can be
altered by any implementor. When overriding a visit_XXX method, the
implementation can decide to continue with the default walking behavior
(by calling the walk function on the passed AstStatement-Struct, to skip it,
or to continue with an alternative walking bahavior.

removed unused AST element CastStatement

lewis-revill committed:
feat: Access references to globals through a custom GOT

This commit introduces the association of GOT indices to the LLVM index,
which then allows us to utilise that when generating references to
variables to check if a given variable has an entry in the GOT. If so,
we obtain its index, and generate the necessary LLVM IR to access the
address contained within the GOT rather than accessing the variable
directly.

ghaith committed:
Merge pull request #1198 from embecosm/move-variables-to-sections

Move variables to ELF sections with mangled names

ghaith committed:
Merge pull request #1199 from embecosm/encode-complex-types

Encode complex types in section mangling

ghaith committed:
Merge pull request #1205 from embecosm/add-decoding-to-section-mangler

Add decoding to section mangler

ghaith committed:
Merge remote-tracking branch 'origin/master' into workflow_fixes

ghaith committed:
Merge pull request #1246 from PLC-lang/workflow_fixes

Renamed the build workflow

volsa committed:
feat: Validate argument count (#1233)

Resolves https://github.com/PLC-lang/rusty/issues/1027 and https://github.com/PLC-lang/rusty/issues/1201

volsa committed:
ci: Introduce llvm-lit for Linux (#1243)

This PR introduces llvm-lit, which at some point will replace our current correctness tests. For now only the functionality was introduced as well as GitHub workflows executing example lit tests.

lewis-revill committed:
feat: Generate calls via the GOT

This change involves moving the generation of the GOT from
variable_generator.rs to codegen.rs, in order to also cover not only
global variables but also functions and 'programs' too. Once these have
been given an associated index in the GOT we can use that to replace
normal direct function calls with indirect calls to a function pointer
stored in the GOT. We don't do this for calls with external linkage
since these won't be subject to online change.

volsa committed:
fix: Bitaccess for VAR_OUTPUT (#1214)

Resolves https://github.com/PLC-lang/rusty/issues/1176

CohenArthur committed:
rusty: Work around race conditions when compiling multiple modules

This commit also adds workarounds to make the integration of the generated
code with our online change runtime work well.

wip: almost done with Mutex hashmap

NOTE: Reuse got_indices member from LlvmIndex instead

wip: it works?

src/test_utils: wip: Make these function compile after latest changes

section_mangler: Fix emitted format

wip

codegen custom GOT as a non-external array

wip: running onlinechangexmpl almost works! crashes on signal1()

more hacks

cleanup

cleanup

CohenArthur committed:
section_mangler: Add test for parsing qualified name in section name

CohenArthur committed:
variable_generator: Use qualified name when mangling variable sections

mhasel committed:
fix: for loops no longer execute once when condition is already met (#1248)

* fix: for loop condition

This PR fixes for loops executing once when the predicate already should not be met for decrementing loops.
I have also re-implemented the codegen logic for for-loops, resulting in fewer predecessors and hopefully more
readable IR.

Resolves #1207

nturley-copia committed:
feat: Add --ast CLI argument to emit the AST to stdout (#1256)

* add ast cli argument to print the AST to stdout

* use pretty printed format

Co-authored-by: Volkan <volkanxvs@gmail.com>

---------

Co-authored-by: Volkan <volkanxvs@gmail.com>

volsa committed:
feat: Introduce `REF=` and `REFERENCE_TO` (#1251)

This PR introduces two new keywords, namely `REF=` and `REFERENCE TO`:
* `REF=` is essentially syntactic sugar for an assignment where the right-hand side is wrapped in a `REF()` function call. Therefore `foo := REF(bar)` and `foo REF= bar` are equivalent.
* `REFERENCE TO` is identical to `REF_TO` with the exception of being auto-deref by default. A variable `foo` declared as `REFERENCE TO` will therefore auto-deref on assignments, i.e. `foo := 5` is equivalent to `foo^ := 5`.

More information on CodeSys' [REF=](https://help.codesys.com/api-content/2/codesys/3.5.12.0/en/_cds_ref_assignment/) and [REFERENCE TO](https://help.codesys.com/api-content/2/codesys/3.5.12.0/en/_cds_datatype_reference/) documentation pages.

volsa committed:
chore: Enable logging in VSCode debug view (#1262)

Prints any log calls in VSCodes debug console

dependabot[bot] committed:
chore: Bump openssl from 0.10.64 to 0.10.66 (#1264)

Bumps [openssl](https://github.com/sfackler/rust-openssl) from 0.10.64 to 0.10.66.
- [Release notes](https://github.com/sfackler/rust-openssl/releases)
- [Commits](https://github.com/sfackler/rust-openssl/compare/openssl-v0.10.64...openssl-v0.10.66)

---
updated-dependencies:
- dependency-name: openssl
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

volsa committed:
feat: Aliasing (#1258)

This commit introduces aliasing, where the `AT` keyword can be used on a reference declaration to create a `REFERENCE TO` pointer. Specifically `foo AT bar : DINT` internally resolves to `foo : REFERENCE TO DINT := REF(bar)`.

volsa committed:
fix: Match Option value when parsing alias variables (#1266)

Fixes an issue where omitting a data-type in an alias variable would cause a panic in the parser. For example

```
FUNCTION main
    VAR
        s AT str; // omitted data-type
    END_VAR
END_FUNCTION
```

volsa committed:
chore: Update README (#1261)

CohenArthur committed:
rusty: Update test snapshots
